src/engine/sync/abletonlink.cpp

```diff
@@ -0,0 +1,248 @@
+#include "engine/sync/abletonlink.h"
+
+#include <QtDebug>
+#include <cmath>
+
+#include "control/controlobject.h"
+#include "control/controlpushbutton.h"
+#include "engine/sync/enginesync.h"
+#include "preferences/usersettings.h"
+#include "util/logger.h"
+
+namespace {
+const mixxx::Logger kLogger("AbletonLink");
+constexpr mixxx::Bpm kDefaultBpm(9999.9);
+} // namespace
+
+AbletonLink::AbletonLink(const QString& group, EngineSync* pEngineSync)
+        : m_group(group),
+          m_pEngineSync(pEngineSync),
+          m_mode(SyncMode::None),
+          m_oldTempo(kDefaultBpm),
+          m_currentLatency(0),
+          m_hostTimeAtStartCallback(0),
+          m_sampleTimeAtStartCallback(0) {
+    m_pLink = std::make_unique<ableton::Link>(120.);
+    m_timeAtStartCallback = m_pLink->clock().micros();
+
+    m_pLinkButton = new ControlPushButton(ConfigKey(group, "sync_enabled"));
+    m_pLinkButton->setButtonMode(ControlPushButton::TOGGLE);
+    m_pLinkButton->setStates(2);
+
+    connect(m_pLinkButton,
+            &ControlObject::valueChanged,
+            this,
+            &AbletonLink::slotControlSyncEnabled,
+            Qt::DirectConnection);
```

Comment:
Why is here a direct connection? This will call the slot form the GUI and the controller thread which may cause a race condition.